File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Common/Link.cs
﻿using System;
using JDI_Commons;
using Epam.JDI.Core.Interfaces.Common;
using JDI_Web.Selenium.Base;
using JDI_Web.Selenium.Elements.Base;
using OpenQA.Selenium;
using RestSharp.Extensions;

namespace JDI_Web.Selenium.Elements.Common
{
    public class Link : ClickableText, ILink
    {
        public Link() : this(null) { }
        public Link(By byLocator = null, IWebElement webElement = null)
            : base(byLocator, webElement) { }

        protected Func<WebBaseElement, string> GetReferenceFunc =
          el => el.WebAvatar.FindImmediately(() => el.WebElement.GetAttribute("href"), "");
        public string GetReference()
        {
            return Invoker.DoJActionResult("Get link reference", GetReferenceFunc, href => $"Get href of link '{href}'");
        }
        
        public string WaitReferenceContains(string text)
        {
            Func<WebBaseElement, Func<string>> textFunc = el => () => GetReferenceFunc(el);
            return Invoker.DoJActionResult(
                    $"Wait link contains '{text}'",
                    el => textFunc(el).GetByCondition(t => t.Contains(text))
                    );
        }

        public string WaitMatchReference(string regEx)
        {
            Func<WebBaseElement, Func<string>> textFunc = el => () => GetReferenceFunc(el);
            return Invoker.DoJActionResult(
                    $"Wait link match regex '{regEx}'",
                    el => textFunc(el).GetByCondition(t => t.Matches(regEx))
                    );
        }

        protected Func<WebBaseElement, string> GetTooltipFunc =
          el => el.WebAvatar.FindImmediately(() => el.WebElement.GetAttribute("title"), "");
        public string GetTooltip()
        {
            return Invoker.DoJActionResult("Get link tooltip", GetTooltipFunc, href => $"Get link tooltip '{href}'");
        }
    }
}

File: Python/JDI/web/selenium/elements/common/link.py
from JDI.core.utils.decorators import scenario
from JDI.web.selenium.elements.base.clickable_text import ClickableText


class Link(ClickableText):

    def __init__(self, by_locator=None):
        super(Link, self).__init__(by_locator)

    @scenario("Get link reference")
    def get_reference(self):
        return self.get_reference_action()

    def get_reference_action(self):
        return self.get_web_element().get_attribute("href")

    @scenario(action_name="Get link tooltip")
    def get_tooltip(self):
        return self.get_tooltip_action()

    def get_tooltip_action(self):
        return self.get_web_element().get_attribute("title")


